Kernel/NumTraits.hpp

```diff
@@ -86,128 +94,314 @@ struct NumTraits;
-                  { IMPL_NUM_TRAITS__ARG_EXPR( arity ) });                                                    \
-    }                                                                                                         \
-
-
-
-#define IMPL_NUM_TRAITS__INTERPRETED_FUN(name, SORT_SHORT, _INTERPRETATION, arity)                            \
-    IMPL_NUM_TRAITS__INTERPRETED_SYMBOL(name, SORT_SHORT, _INTERPRETATION)                                    \
-                                                                                                              \
-    static TermList name(IMPL_NUM_TRAITS__ARG_DECL(TermList, arity)) {                                        \
-      return TermList(                                                                                        \
-          Term::create(                                                                                       \
-            name##F(),                                                                                        \
-            { IMPL_NUM_TRAITS__ARG_EXPR(arity) }));                                                           \
-    }                                                                                                         \
-
-#define IMPL_NUM_TRAITS__SPECIAL_CONSTANT(name, value, isName)                                                \
-    constexpr static ConstantType name ## C = ConstantType(value);                                            \
-    static Term* name ## T() {  /* TODO refactor to const& Term */                                            \
-      static Term* trm = theory->representConstant(name ## C);                                                \
-      return trm;                                                                                             \
-    }                                                                                                         \
-    static TermList name()                                                                                    \
-    { return TermList(name ## T()); }                                                                         \
-                                                                                                              \
-    static bool isName(const TermList& l)                                                                     \
-    { return l == name(); }                                                                                   \
-
-#define IMPL_NUM_TRAITS__QUOTIENT_REMAINDER(SHORT, X)                                                         \
-    IMPL_NUM_TRAITS__INTERPRETED_FUN( quotient ## X, SHORT,  _QUOTIENT_ ## X, 2)                              \
-    IMPL_NUM_TRAITS__INTERPRETED_FUN(remainder ## X, SHORT, _REMAINDER_ ## X, 2)                              \
+#define IMPL_NUM_TRAITS__INTERPRETED_SYMBOL(name, Name, SORT_SHORT, _INTERPRETATION)      \
```

Comment:
I'm going to assume that you know what you're doing here. I think a lot of the diff is just introducing uppercase `Name`, right?

Reply:
Yes and also adding quite a bit of misc little helper functions to be able to query theory function symbols more nicel.